runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java

```diff
@@ -0,0 +1,601 @@
+        context.getDataSetOrThrow(Iterables.getOnlyElement(transform.getInputsMap().values()));
+    context.addDataSet(
+        Iterables.getOnlyElement(transform.getOutputsMap().values()), inputDataSet.rebalance());
+  }
+
+  private static <T> void translateAssignWindows(
+      String id, RunnerApi.Pipeline pipeline, BatchTranslationContext context) {
+    RunnerApi.Components components = pipeline.getComponents();
+    RunnerApi.PTransform transform = components.getTransformsOrThrow(id);
+    RunnerApi.WindowIntoPayload payload;
+    try {
+      payload = RunnerApi.WindowIntoPayload.parseFrom(transform.getSpec().getPayload());
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalArgumentException(e);
+    }
+    WindowFn<T, ? extends BoundedWindow> windowFn =
+        (WindowFn<T, ? extends BoundedWindow>)
+            WindowingStrategyTranslation.windowFnFromProto(payload.getWindowFn());
+
+    String inputCollectionId = Iterables.getOnlyElement(transform.getInputsMap().values());
+    String outputCollectionId = Iterables.getOnlyElement(transform.getOutputsMap().values());
+    Coder<WindowedValue<T>> outputCoder = instantiateCoder(outputCollectionId, components);
+    TypeInformation<WindowedValue<T>> resultTypeInfo = new CoderTypeInformation<>(outputCoder);
+
+    DataSet<WindowedValue<T>> inputDataSet = context.getDataSetOrThrow(inputCollectionId);
+
+    FlinkAssignWindows<T, ? extends BoundedWindow> assignWindowsFunction =
+        new FlinkAssignWindows<>(windowFn);
+
+    DataSet<WindowedValue<T>> resultDataSet =
+        inputDataSet
+            .flatMap(assignWindowsFunction)
+            .name(transform.getUniqueName())
+            .returns(resultTypeInfo);
+
+    context.addDataSet(outputCollectionId, resultDataSet);
+  }
+
+  private static <InputT> void translateExecutableStage(
+      String id, RunnerApi.Pipeline pipeline, BatchTranslationContext context) {
+    // TODO: Fail on stateful DoFns for now.
+    // TODO: Support stateful DoFns by inserting group-by-keys where necessary.
+    // TODO: Fail on splittable DoFns.
+    // TODO: Special-case single outputs to avoid multiplexing PCollections.
+
+    RunnerApi.Components components = pipeline.getComponents();
+    RunnerApi.PTransform transform = components.getTransformsOrThrow(id);
+    Map<String, String> outputs = transform.getOutputsMap();
+    // Mapping from PCollection id to coder tag id.
+    BiMap<String, Integer> outputMap = createOutputMap(outputs.values());
+    // Collect all output Coders and create a UnionCoder for our tagged outputs.
+    List<Coder<?>> unionCoders = Lists.newArrayList();
+    // Enforce tuple tag sorting by union tag index.
+    Map<String, Coder<WindowedValue<?>>> outputCoders = Maps.newHashMap();
+    for (String collectionId : new TreeMap<>(outputMap.inverse()).values()) {
+      Coder<WindowedValue<?>> windowCoder = (Coder) instantiateCoder(collectionId, components);
```

Comment:
<!--new_thread; commit:c3bc9f1ff4d5f830b19aef83b758b12e363f7679; resolved:0-->
Could you give this a more descriptive name rather than "instantiateCoder"?